crates/hotshot-stake-table/src/vec_based.rs

```diff
@@ -0,0 +1,495 @@
+    }
+
+    fn lookup(
+        &self,
+        version: SnapshotVersion,
+        key: &Self::Key,
+    ) -> Result<Self::Amount, StakeTableError> {
+        let table = self.get_version(version)?;
+        let pos = self.lookup_pos(key)?;
+        if pos >= table.bls_keys.len() {
+            Err(StakeTableError::KeyNotFound)
+        } else {
+            Ok(table.stake_amount[pos])
+        }
+    }
+
+    fn lookup_with_proof(
+        &self,
+        version: SnapshotVersion,
+        key: &Self::Key,
+    ) -> Result<(Self::Amount, Self::LookupProof), StakeTableError> {
+        let amount = self.lookup(version, key)?;
+        Ok((amount, ()))
+    }
+
+    fn lookup_with_aux_and_proof(
+        &self,
+        version: SnapshotVersion,
+        key: &Self::Key,
+    ) -> Result<(Self::Amount, Self::Aux, Self::LookupProof), StakeTableError> {
+        let table = self.get_version(version)?;
+        let pos = self.lookup_pos(key)?;
+        if pos >= table.bls_keys.len() {
+            Err(StakeTableError::KeyNotFound)
+        } else {
+            Ok((table.stake_amount[pos], table.schnorr_keys[pos].clone(), ()))
+        }
+    }
+
+    fn update(
+        &mut self,
+        key: &Self::Key,
+        delta: Self::Amount,
+        negative: bool,
+    ) -> Result<Self::Amount, StakeTableError> {
+        let pos = self.lookup_pos(key)?;
+        if negative {
+            if delta > self.head.stake_amount[pos] {
+                return Err(StakeTableError::InsufficientFund);
+            }
+            self.head_total_stake -= delta;
+            self.head.stake_amount[pos] -= delta;
+        } else {
+            self.head_total_stake += delta;
+            self.head.stake_amount[pos] += delta;
+        }
+        Ok(self.head.stake_amount[pos])
+    }
+
+    fn sample(
+        &self,
+        rng: &mut (impl SeedableRng + CryptoRngCore),
+    ) -> Option<(&Self::Key, &Self::Amount)> {
+        let mut bytes = [0u8; 64];
+        rng.fill_bytes(&mut bytes);
+        let r = U512::from_big_endian(&bytes);
+        let m = U512::from(self.last_epoch_start_total_stake);
+        let mut pos: U256 = (r % m).try_into().unwrap(); // won't fail
+        let idx = 0;
+        while pos > self.last_epoch_start.stake_amount[idx] {
+            pos -= self.last_epoch_start.stake_amount[idx];
+        }
+        Some((
+            &self.last_epoch_start.bls_keys[idx],
+            &self.last_epoch_start.stake_amount[idx],
+        ))
+    }
+
+    fn try_iter(&self, version: SnapshotVersion) -> Result<Self::IntoIter, StakeTableError> {
+        let table = self.get_version(version)?;
+        let owned = (0..table.bls_keys.len())
+            .map(|i| {
+                (
+                    table.bls_keys[i].clone(),
+                    table.stake_amount[i],
+                    table.schnorr_keys[i].clone(),
+                )
+            })
+            .collect::<Vec<_>>();
+        Ok(owned.into_iter())
```

Comment:
this looks slightly strange to me, why taking an iterator, collect it, and then into an iterator again?
I think we can directly return `(0..table.bls_keys.len()).map()` (or maybe adding some `.to_owned()` or `.into_iter()` in the middle to make the compiler happy?

Reply:
That's the problem, I tried but couldn't make the compiler happy :(

Reply:
an idea: change the `Self::IntoIter = Zip<Zip<Self::Key, Self::Amount>, Self::Aux>`, then you can just return `table.bls_keys.iter().zip(table.stake_amount.iter()).zip(table.schnorr_keys.iter())`?

Reply:
Nah, you are missing lifetime specifiers.